### PR TITLE
feat: Add create/update ui for scheduler_opts

### PR DESCRIPTION
**Pull request by ggstone0523**
**Resolve: lablup/backend.ai-webui #1190**

- display scheduler_opts's each key and value to table.
PR6_review0_1

- scheduler_opts UI in scaling group create operation dialog.
PR6_review0_2

- scheduler_opts UI in scaling group update operation dialog.
PR6_review0_3

- check scheduler_opts's key-value valid
PR6_review0_4

### CODE REVIEW

**Comment by ggstone0523** [comment]
### The revision was completed.

**feature**

- Add dialog for scheduler_opts and driver_opts
PR6_review1_1

- Change set scheduler options dialog
PR6_review1_2

- validation check
PR6_review1_3

**bugfix**

- fix allowed_session_value didn't change

File: src/lib/backend.ai-client-node.ts
Comment (lizable) [inline src/lib/backend.ai-client-node.ts]: This might be a typo error, perhaps?

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: How about using UI layout based on "keypair detail dialog"?
And please refer to key, value based UI Layout [here](https://design.mindsphere.io/patterns/key-value.html).

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Please check a CSS variable that manages the height of `mwc-list-item`, instead of using inline height in every HTML element.
You can see in [Styled section in README file of mwc-select](https://github.com/material-components/material-web/tree/master/packages/select).

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: I'm not sure about using "nested dialog" here. Could you explain the reason for this layout?

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Any reasons for using `_clearValues()`?
If you want to restore the values to the previous value, then we should change the name of functions to more proper terms.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Three things to ask in this section:
1. I'm not sure `pending timeout` is the required value.
2. And Let's replace `_(underbar)` in the label to whitespace since is a bit unnatural.
3. Let's use change event instead of input event, since `input event` is triggered much more frequently compared to "change event". Please refer to [this link](https://stackoverflow.com/questions/17047497/difference-between-change-and-input-event-for-an-input-element).

**Review by lizable — CHANGES_REQUESTED** [review]
I have several things to request. Please check out the comments below.
Some of them need your explanations.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: Yes, 
 As the scheduler option is added, a new input form for the option should be added, and if these input forms are not separated through dialog, the length of the create/modify dialog will become longer and I think less readable each time the option is added.

Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: Hmm, but the number of keys may not exceed 5 or 6, so I would like to suggest accordion UI instead of nested dialog.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: because to reset scheduler options's value to default when user open 'create dialog'.
and i changed name '_clearValues' to '_resetAllValues'

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: 1. The reason why I put pending_timeout was because pending_timeout came into thisPR(https://github.com/lablup/backend.ai-manager/pull/527) as a new scheduler_opts.
2. I revise it following comment.
3. I revise it following comment.

File: src/lib/backend.ai-client-node.ts
Comment (ggstone0523) [reply src/lib/backend.ai-client-node.ts]: yes it was a typo error, so i revise it to 'scheduler_opts'

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: I revise it following comment.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: I revise it following 'Key above value' key value pair types.

**image**
PR6_review2_1

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: I revise it following comment.

**image**
PR6_review2_2

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: Please check again about "keypair detail dialog" and update the UI according to that.

**Comment by ggstone0523** [comment]
### The revision was completed.

1. change resource group detail dialog title name from 'Detailed Information' to 'Resource Group Detail'
and change resource group detail dialog UI according to keypair detail dialog UI.

**image**
PR6_review3_1

2. prevent create/modify dialog scroll 
and check that if input is maximum length the description component display all input.

**image**
PR6_review3_2
PR6_review3_3

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Why would we need to use "text" type in the input field that only allows "number"?
It would bring redundant validation which can be handled by just assigning the type of input field to "number".
Please check the README file in `mwc-textfield` and set it with `min` value.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Could you change this function without any args?
I guess we can assign the value inside the function.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Please remove args, It doesn't seem to be "essential" in here.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Please remove the args in this function, too.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's not use magic number in here ("2"), you need to change that number to more descriptive variable, instead.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Any reason for adding args in this way?

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: This function also needs to be reconsidered when you change the type of corresponding input field from "text" to "number".

Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: For now, I don't think it's necessary.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's remove this function since we don't provide modification of "driver option" yet.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's create a wrapping function for better readability.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: And also I found there's a typo error in the function name. Please change it to `SchedulerOpts` instead of using `SchulerOpts`.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: Also, I'm not sure about doing extra operation in "save" function. 
Could we just call `verifySchedulerOpts` outside in the save function so that save function only called when all value is valid?
It seems this function does too many things inside compared to it's name.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's use the option as a component property instead of using raw text.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Is this element id supposed to have "keypair-detail-dialog"?

Comment (lizable) [reply src/components/backend-ai-scaling-group-list.ts]: And most of detail dialog contains not only for information that didn't cover in table UI, but it also contains other essential information. Please check out user detail dialog or vfolder detail dialog. You may get a clue from there.

**Review by lizable — CHANGES_REQUESTED** [review]
@ggstone0523 
Before you ask for a PR review, I have a few things to request.
- Please check the "typo" and fix it.
- Please avoid using extra args in function unless it's necessary.
- Please create the function and element ID with a certain pattern for better readability.
- If you are struggling with the layouts, referring to other dialog would be welcomed.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts:765]: I revise it following comment.

Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts:765]: I revise it following comment.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts:591]: I revise it following comment.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts:579]: I revise it following comment.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: I revise it following comment.

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: I revise it following comment.

**Comment by ggstone0523** [comment]
### The revision was completed by reflecting the comments.

- Using min value in mwc-textfield for validate scheduler opts input forms value
- Remove unnecessary argument in function
- Divide save function to verify function and save function
- Change `schulerOpts` to `schedulerOpts`
- Using Array.prototype.includes() to remove magic number
- Change way that adding args to _setSchedulerOptsInputForm bacause improve readability
- Remove pending timeout validator
- Remove drive option renderer
- Add wrapping function for add button click event
- replace raw text to property in 'allowed session types' scheduler opts input forms
- change resource-group-detail-dialog UI
**image**
PR6_review4_1

File: src/components/backend-ai-scaling-group-list.ts
Comment (ggstone0523) [reply src/components/backend-ai-scaling-group-list.ts]: because i change 'object in object' type to 'list in list' type to use map.
so, i approach key and value using list index. 
but, this approach is reduce readablity. so i change this logic

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's wrap it with a function instead of stacking it all in the event handler.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Same in here.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Could we just remove the extra renderer(_schedulerOptionsRenderer) and put all of HTML into render function? I think it's too much to display a single dialog information.

File: src/components/backend-ai-scaling-group-list.ts
Comment (lizable) [inline src/components/backend-ai-scaling-group-list.ts]: Let's not use another condition statement in switch statement.

**Review by lizable — CHANGES_REQUESTED** [review]

**Comment by ggstone0523** [comment]
### The revision was completed by reflecting the comments.

- apply wrapping function for control panel buttons
- remove switch statement in '_initializeModifySchedulerOpts' function and change it to if statement
- remove '_schedulerOptionsRenderer' function and put all to render function
- remove required in pending_timeout input form

**Review by lizable — COMMENTED** [review]
I fixed a few minor linting errors and got rid of some legacy codes.
We need a review and approval for this PR from one of the maintainers, @inureyes or @adrysn.

File: src/lib/backend.ai-client-node.ts
Comment (inureyes) [inline src/lib/backend.ai-client-node.ts]: The "order" of parameter should be kept as original (for API compatibility / legacy support). scheduler_opts will be the end of parameter.

File: src/lib/backend.ai-client-node.js
Comment (inureyes) [inline src/lib/backend.ai-client-node.js]: The "order" of parameter should be kept as original (for API compatibility / legacy support). scheduler_opts will be the end of parameter.

**Review by inureyes — CHANGES_REQUESTED** [review]
Check the comments and related calls on backend-ai-scaling-group-list.ts

**Comment by ggstone0523** [comment]
### The revision was completed by reflecting the comments.

- change order of create function's parameter (scheduler_opts go to end of parameter)

**Comment by lizable** [comment]
I changed several things in order to keep the version compatibility and also the identity of terms used, especially in "scaling group".

- any `scaling group` terms used in WebUI component will be changed to `resource group`.
- since the milestone of the related issue([backend.ai#343](https://github.com/lablup/backend.ai/issues/343)) in Backend.AI component is `22.03`, configuring scheduler option will be supported from `22.03`.

**Review by inureyes — APPROVED** [review]
Goot to go

**Pull request merged**
